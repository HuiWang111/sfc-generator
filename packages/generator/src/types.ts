import type generate from '@babel/generator'
import type { GeneratorOptions } from '@babel/generator'
import type { SFCBlock } from '@vue/compiler-sfc'
import type { TemplateNodeType } from './constants'

export interface TemplateNodeAttr {
  name: string
  value: string
  dynamic?: boolean | undefined
}

export interface TemplateNodeDirective {
  name: string
  rawName: string
  value: string
  arg: null
  isDynamicArg: false
  modifiers: null
}

export interface TemplateNodeModel {
  value: string
  expression: string // eg: '"value"'
  callback: string
}

export type TemplateNode =
  | ElementNode
  | InterpolationNode
  | TextNode

export interface ElementNode {
  type: TemplateNodeType.Element
  tag: string
  attrs?: TemplateNodeAttr[]
  attrsList?: Omit<TemplateNodeAttr, 'dynamic'>[]
  attrsMap?: Record<string, string> // 包含 events / class
  rawAttrsMap?: Record<string, string>
  parent?: TemplateNode | undefined
  children: TemplateNode[]
  plain?: boolean
  static?: boolean
  staticRoot?: boolean
  events?: Record<string, {
    value: string
    dynamic: boolean
  }>
  staticClass?: string // eg: '"submit-container"'
  hasBindings?: boolean
  forProcessed?: boolean
  alias?: string // v-for 定义的变量，eg: v-for="data in list" 中的 data
  for?: string // 被循环的变量，eg: v-for="data in list" 中的 list
  directives?: TemplateNodeDirective[]
  model?: TemplateNodeModel
  __skip__?: boolean
}

export interface InterpolationNode {
  type: TemplateNodeType.Interpolation
  tokens: object[]
  expression: string
  text: string
  parent?: TemplateNode | undefined
  __skip__?: boolean
}

export interface TextNode {
  type: TemplateNodeType.Text
  text: string
  parent?: TemplateNode | undefined
  __skip__?: boolean
}

export interface GenerateOptions {
  breakOnAttrs?: number
  indentSize?: number
  autoIndent?: boolean
}

export type BabelNode = Parameters<typeof generate>[0]

export interface ScriptAttrs {
  setup?: boolean
  lang?: string
}

export interface SFCGenerateOptions {
  template: {
    node: TemplateNode
    options?: GenerateOptions
  }
  script?: {
    node: BabelNode
    attrs?: ScriptAttrs
    options?: GeneratorOptions
  }
  styles?: SFCBlock[]
}

export interface ScriptParseOptions<T = boolean> {
  setup?: T
  lang?: 'ts' | 'js'
  jsx?: boolean
}

export interface OptionOperator {
  add: (...args: any[]) => any
  update: (...args: any[]) => any
  remove: (name: string) => any
  get: (name: string) => any
}
