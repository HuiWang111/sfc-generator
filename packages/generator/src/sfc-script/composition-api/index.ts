import type { Statement } from '@babel/types'
import {
  ComputedOption,
  PropsOption,
  RefOption,
} from './properties'

export class CompositionApi {
  refOption: RefOption | null = null
  computedOption: ComputedOption | null = null
  propsOption: PropsOption | null = null

  constructor(private statements: Statement[]) {}

  data() {
    if (!this.refOption)
      this.refOption = new RefOption(this.statements)
    return this.refOption
  }

  computed() {
    if (!this.computedOption)
      this.computedOption = new ComputedOption(this.statements)
    return this.computedOption
  }

  props() {
    if (!this.propsOption)
      this.propsOption = new PropsOption(this.statements)
    return this.propsOption
  }
}
