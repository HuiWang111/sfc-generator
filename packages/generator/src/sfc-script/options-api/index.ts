import type { NodePath } from '@babel/traverse'
import type { ObjectExpression, ObjectMethod, ObjectProperty } from '@babel/types'
import * as t from '@babel/types'
import {
  ComputedOption,
  DataOption,
  MethodsOption,
  PropsOption,
  WatchOption,
} from './properties'

export class OptionsApi {
  // private isSetup: boolean
  private dataOption: DataOption | null = null
  private computedOption: ComputedOption | null = null
  private watchOption: WatchOption | null = null
  private methodsOption: MethodsOption | null = null
  private propsOption: PropsOption | null = null

  constructor(private _node: NodePath<ObjectExpression>) {}

  private getOptionNode<T extends ObjectMethod | ObjectProperty>(name: string) {
    return this._node.get('properties')
      .find((prop): prop is NodePath<T> => {
        const keyNode = prop.get('key')
        return keyNode.isIdentifier() && keyNode.node.name === name
      })
  }

  get node() {
    return this._node
  }

  data() {
    if (this.dataOption) {
      return this.dataOption
    }

    const dataNode = this.getOptionNode<ObjectMethod>('data')
    if (dataNode) {
      this.dataOption = new DataOption(dataNode.node, this)
      return this.dataOption
    }

    const objectMethod = t.objectMethod(
      'method',
      t.identifier('data'),
      [],
      t.blockStatement([
        t.returnStatement(
          t.objectExpression([]),
        ),
      ]),
    )
    this._node.pushContainer('properties', objectMethod)
    this.dataOption = new DataOption(objectMethod, this)
    return this.dataOption
  }

  computed() {
    if (this.computedOption)
      return this.computedOption

    const computedNode = this.getOptionNode<ObjectProperty>('computed')
    if (computedNode) {
      this.computedOption = new ComputedOption(computedNode.node, this)
      return this.computedOption
    }

    const objectProperty = t.objectProperty(
      t.identifier('computed'),
      t.objectExpression([]),
    )
    this._node.pushContainer('properties', objectProperty)
    this.computedOption = new ComputedOption(objectProperty, this)
    return this.computedOption
  }

  watch() {
    if (this.watchOption)
      return this.watchOption

    const watchNode = this.getOptionNode<ObjectProperty>('watch')
    if (watchNode) {
      this.watchOption = new WatchOption(watchNode.node, this)
      return this.watchOption
    }

    const objectProperty = t.objectProperty(
      t.identifier('watch'),
      t.objectExpression([]),
    )
    this._node.pushContainer('properties', objectProperty)
    this.watchOption = new WatchOption(objectProperty, this)
    return this.watchOption
  }

  methods() {
    if (this.methodsOption)
      return this.methodsOption

    const methodsNode = this.getOptionNode<ObjectProperty>('methods')
    if (methodsNode) {
      this.methodsOption = new MethodsOption(methodsNode.node, this)
      return this.methodsOption
    }

    const objectProperty = t.objectProperty(
      t.identifier('methods'),
      t.objectExpression([]),
    )
    this._node.pushContainer('properties', objectProperty)
    this.methodsOption = new MethodsOption(objectProperty, this)
    return this.methodsOption
  }

  props() {
    if (this.propsOption)
      return this.propsOption

    const propsNode = this.getOptionNode<ObjectProperty>('props')
    if (propsNode) {
      this.propsOption = new PropsOption(propsNode.node, this)
      return this.propsOption
    }

    const objectProperty = t.objectProperty(
      t.identifier('props'),
      t.objectExpression([]),
    )
    this._node.pushContainer('properties', objectProperty)
    this.propsOption = new PropsOption(objectProperty, this)
    return this.propsOption
  }
}
