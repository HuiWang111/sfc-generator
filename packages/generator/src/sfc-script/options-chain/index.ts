export { ObjectExpressionOptionsChain } from './object-expression'
