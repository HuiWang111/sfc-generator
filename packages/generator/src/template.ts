import type { ElementNode, GenerateOptions, TemplateNode } from './types'
import { TemplateNodeType } from './types'

export class Template {
  constructor(private options: GenerateOptions = {}) {}

  private padCode(str: string, indentCount: number) {
    const { indent = 2 } = this.options
    const spaceCount = indent * indentCount
    return str.padStart(str.length + spaceCount)
  }

  private generateAttrCode(
    node: ElementNode,
    indentCount: number,
  ): [string, boolean] {
    const { breakOnAttrs = 2 } = this.options
    const attrEntries = Object.entries(node.attrsMap)
    if (!attrEntries.length) {
      return ['', false]
    }

    const shouldBreak = attrEntries.length > breakOnAttrs
    const attrCodes = (
      attrEntries
        .map(([key, value]) => {
          if (shouldBreak) {
            return this.padCode(`${key}="${value}"`, indentCount)
          }
          return `${key}="${value}"`
        })
    )

    if (!shouldBreak) {
      return [attrCodes.join(' '), shouldBreak]
    }
    return [attrCodes.join('\n'), shouldBreak]
  }

  private generateEndTag(
    node: ElementNode,
    indentCount: number,
    shouldBreak: boolean,
  ) {
    if (!node.children || !node.children.length) {
      return this.padCode('/>', indentCount)
    }
    const childrenCode = (
      node.children
        .map((c) => {
          return this.generate(c, indentCount + 1)
        })
        .join('')
    )
    return [
      this.padCode('>', shouldBreak ? indentCount : 0),
      childrenCode,
      this.padCode(`</${node.tag}>`, indentCount),
    ].join('\n')
  }

  generate(node: TemplateNode, indentCount = 1): string {
    if (
      node.type === TemplateNodeType.Interpolation
      || node.type === TemplateNodeType.Text
    ) {
      if (node.type === TemplateNodeType.Text && node.text === ' ') {
        return '\n'
      }
      return this.padCode(node.text, indentCount)
    }

    const [attrCode, shouldBreak] = this.generateAttrCode(node, indentCount + 1)
    const startTag = this.padCode(`<${node.tag}`, indentCount)
    const endTag = this.generateEndTag(node, indentCount, shouldBreak)

    if (!shouldBreak) {
      return [
        startTag,
        attrCode ? ' ' : '',
        attrCode,
        endTag,
      ]
        .join('')
    }
    return [startTag, attrCode, endTag].join('\n')
  }
}
