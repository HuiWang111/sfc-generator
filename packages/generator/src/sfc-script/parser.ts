import type { ParserOptions } from '@babel/parser'
import type { ScriptParseOptions } from '../types'
import { parse as babelParse } from '@babel/parser'
import traverse from '@babel/traverse'
import { CompositionApi } from './composition-api'
import { OptionsApi } from './options-api'

export function parse<T extends boolean = false>(
  code: string,
  options: ScriptParseOptions<T> = {},
): {
    api: T extends true ? CompositionApi : OptionsApi
    ast: ReturnType<typeof babelParse>
  } {
  const { lang, jsx = false, setup = false } = options

  const plugins: ParserOptions['plugins'] = []
  if (lang === 'ts') {
    plugins.push('typescript')
  }
  if (jsx) {
    plugins.push('jsx')
  }

  const ast = babelParse(code, {
    sourceType: 'module',
    plugins,
  })

  let api: OptionsApi | CompositionApi | null = null

  if (setup) {
    api = new CompositionApi(ast.program.body)
  }
  else {
    traverse(ast, {
      ExportDefaultDeclaration(path) {
        if (setup)
          return

        const declarationPath = path.get('declaration')
        if (declarationPath.isObjectExpression()) {
          api = new OptionsApi(declarationPath)
        }
      },
    })
  }

  if (!api) {
    throw new Error(`script setup attr is ${setup}, parse code error: ${code}`)
  }

  return {
    api: api as any,
    ast,
  }
}
