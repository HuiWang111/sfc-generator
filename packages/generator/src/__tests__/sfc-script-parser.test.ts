import generate from '@babel/generator'
import template from '@babel/template'
import * as t from '@babel/types'
import { parse } from '../sfc-script'

describe('sfc script parser should work', () => {
  it('options update data should work', () => {
    const js = `export default {
  name: 'AInput',
  data () {
    return {
      value: '1'
    }
  },
}`

    const { ast, api } = parse(js)
    if (!api)
      return

    api.data().add('number', t.numericLiteral(1))
    let { code } = generate(ast)
    expect(code).toMatchSnapshot()

    api.data().update('value', t.booleanLiteral(true));
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()

    api.data().remove('value');
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()

    api.data().remove('number');
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()
  })

  it('options create data should work', () => {
    const js = `export default {
  name: 'AInput',
}`

    const { ast, api } = parse(js)
    if (!api)
      return

    api.data().add('number', t.numericLiteral(1))
    let { code } = generate(ast)
    expect(code).toMatchSnapshot()

    const value = api.data().get('number')
    expect(value).toMatchSnapshot()

    api.data().remove('number');
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()
  })

  it('options update computed should work', () => {
    const js = `export default {
  name: 'AInput',
  data() {
    return {
      firstName: 'Davide',
      lastName: 'Li',
    }
  },
  computed: {
    fullName() {
      return this.firstName + ' ' + this.lastName
    },
  },
}`

    const { ast, api } = parse(js)
    if (!api)
      return

    const returnStatement = template.statement('return this.lastName + \' \' + this.firstName')

    api.computed().add(
      t.objectMethod(
        'method',
        t.identifier('reverseName'),
        [],
        t.blockStatement([
          returnStatement(),
        ]),
      ),
    )
    let { code } = generate(ast)
    expect(code).toMatchSnapshot()

    api.computed().update(
      t.objectMethod(
        'method',
        t.identifier('reverseName'),
        [],
        t.blockStatement([
          template.statement('return this.lastName + this.firstName')(),
        ]),
      ),
    );
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()

    api.computed().remove('reverseName');
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()

    const value = api.computed().get('fullName')
    expect(value).toMatchSnapshot()
  })

  it('options create computed should work', () => {
    const js = `export default {
  name: 'AInput',
}`

    const { ast, api } = parse(js)
    if (!api)
      return

    api.computed().add(
      t.objectMethod(
        'method',
        t.identifier('name'),
        [],
        t.blockStatement([
          template.statement('return this._name')(),
        ]),
      ),
    )
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })

  it('options update watch should work', () => {
    const js = `export default {
  name: 'AInput',
  data() {
    return {
      firstName: 'Davide',
      lastName: 'Li',
      fullName: '',
    }
  },
  watch: {
    firstName(val) {
      this.fullName = val + this.lastName
    }
  },
}`

    const { ast, api } = parse(js)
    if (!api)
      return

    api.watch().add(
      t.objectMethod(
        'method',
        t.identifier('lastName'),
        [t.identifier('val')],
        t.blockStatement([
          template.statement('this.fullName =  this.firstName + \' \' + val;')(),
        ]),
      ),
    )
    let { code } = generate(ast)
    expect(code).toMatchSnapshot()

    api.watch().update(
      t.objectMethod(
        'method',
        t.identifier('firstName'),
        [t.identifier('val')],
        t.blockStatement([
          template.statement('this.fullName = val + \' \' + this.lastName')(),
        ]),
      ),
    );
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()

    api.watch().remove('firstName');
    ({ code } = generate(ast))
    expect(code).toMatchSnapshot()

    const value = api.watch().get('lastName')
    expect(value).toMatchSnapshot()
  })

  it('options create watch should work', () => {
    const js = `export default {
  name: 'AInput',
}`

    const { ast, api } = parse(js)
    if (!api)
      return

    api.watch().add(
      t.objectMethod(
        'method',
        t.identifier('firstName'),
        [t.identifier('val')],
        t.blockStatement([
          template.statement('this.fullName = val + \' \' + this.lastName')(),
        ]),
      ),
    )
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })
})
