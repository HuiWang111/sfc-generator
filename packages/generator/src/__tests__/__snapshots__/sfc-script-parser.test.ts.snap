// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sfc script parser should work options create computed should work 1`] = `
"export default {
  name: 'AInput',
  computed: {
    name() {
      return this._name;
    }
  }
};"
`;

exports[`sfc script parser should work options create data should work 1`] = `
"export default {
  name: 'AInput',
  data() {
    return {
      number: 1
    };
  }
};"
`;

exports[`sfc script parser should work options create data should work 2`] = `
{
  "type": "NumericLiteral",
  "value": 1,
}
`;

exports[`sfc script parser should work options create data should work 3`] = `
"export default {
  name: 'AInput',
  data() {
    return {};
  }
};"
`;

exports[`sfc script parser should work options create watch should work 1`] = `
"export default {
  name: 'AInput',
  watch: {
    firstName(val) {
      this.fullName = val + ' ' + this.lastName;
    }
  }
};"
`;

exports[`sfc script parser should work options update computed should work 1`] = `
"export default {
  name: 'AInput',
  data() {
    return {
      firstName: 'Davide',
      lastName: 'Li'
    };
  },
  computed: {
    fullName() {
      return this.firstName + ' ' + this.lastName;
    },
    reverseName() {
      return this.lastName + ' ' + this.firstName;
    }
  }
};"
`;

exports[`sfc script parser should work options update computed should work 2`] = `
"export default {
  name: 'AInput',
  data() {
    return {
      firstName: 'Davide',
      lastName: 'Li'
    };
  },
  computed: {
    fullName() {
      return this.firstName + ' ' + this.lastName;
    },
    reverseName() {
      return this.lastName + this.firstName;
    }
  }
};"
`;

exports[`sfc script parser should work options update computed should work 3`] = `
"export default {
  name: 'AInput',
  data() {
    return {
      firstName: 'Davide',
      lastName: 'Li'
    };
  },
  computed: {
    fullName() {
      return this.firstName + ' ' + this.lastName;
    }
  }
};"
`;

exports[`sfc script parser should work options update computed should work 4`] = `
Node {
  "async": false,
  "body": Node {
    "body": [
      Node {
        "argument": Node {
          "end": 199,
          "left": Node {
            "end": 183,
            "left": Node {
              "computed": false,
              "end": 177,
              "loc": SourceLocation {
                "end": Position {
                  "column": 27,
                  "index": 177,
                  "line": 11,
                },
                "filename": undefined,
                "identifierName": undefined,
                "start": Position {
                  "column": 13,
                  "index": 163,
                  "line": 11,
                },
              },
              "object": Node {
                "end": 167,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 17,
                    "index": 167,
                    "line": 11,
                  },
                  "filename": undefined,
                  "identifierName": undefined,
                  "start": Position {
                    "column": 13,
                    "index": 163,
                    "line": 11,
                  },
                },
                "start": 163,
                "type": "ThisExpression",
              },
              "property": Node {
                "end": 177,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 27,
                    "index": 177,
                    "line": 11,
                  },
                  "filename": undefined,
                  "identifierName": "firstName",
                  "start": Position {
                    "column": 18,
                    "index": 168,
                    "line": 11,
                  },
                },
                "name": "firstName",
                "start": 168,
                "type": "Identifier",
              },
              "start": 163,
              "type": "MemberExpression",
            },
            "loc": SourceLocation {
              "end": Position {
                "column": 33,
                "index": 183,
                "line": 11,
              },
              "filename": undefined,
              "identifierName": undefined,
              "start": Position {
                "column": 13,
                "index": 163,
                "line": 11,
              },
            },
            "operator": "+",
            "right": Node {
              "end": 183,
              "extra": {
                "raw": "' '",
                "rawValue": " ",
              },
              "loc": SourceLocation {
                "end": Position {
                  "column": 33,
                  "index": 183,
                  "line": 11,
                },
                "filename": undefined,
                "identifierName": undefined,
                "start": Position {
                  "column": 30,
                  "index": 180,
                  "line": 11,
                },
              },
              "start": 180,
              "type": "StringLiteral",
              "value": " ",
            },
            "start": 163,
            "type": "BinaryExpression",
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 49,
              "index": 199,
              "line": 11,
            },
            "filename": undefined,
            "identifierName": undefined,
            "start": Position {
              "column": 13,
              "index": 163,
              "line": 11,
            },
          },
          "operator": "+",
          "right": Node {
            "computed": false,
            "end": 199,
            "loc": SourceLocation {
              "end": Position {
                "column": 49,
                "index": 199,
                "line": 11,
              },
              "filename": undefined,
              "identifierName": undefined,
              "start": Position {
                "column": 36,
                "index": 186,
                "line": 11,
              },
            },
            "object": Node {
              "end": 190,
              "loc": SourceLocation {
                "end": Position {
                  "column": 40,
                  "index": 190,
                  "line": 11,
                },
                "filename": undefined,
                "identifierName": undefined,
                "start": Position {
                  "column": 36,
                  "index": 186,
                  "line": 11,
                },
              },
              "start": 186,
              "type": "ThisExpression",
            },
            "property": Node {
              "end": 199,
              "loc": SourceLocation {
                "end": Position {
                  "column": 49,
                  "index": 199,
                  "line": 11,
                },
                "filename": undefined,
                "identifierName": "lastName",
                "start": Position {
                  "column": 41,
                  "index": 191,
                  "line": 11,
                },
              },
              "name": "lastName",
              "start": 191,
              "type": "Identifier",
            },
            "start": 186,
            "type": "MemberExpression",
          },
          "start": 163,
          "type": "BinaryExpression",
        },
        "end": 199,
        "loc": SourceLocation {
          "end": Position {
            "column": 49,
            "index": 199,
            "line": 11,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 6,
            "index": 156,
            "line": 11,
          },
        },
        "start": 156,
        "type": "ReturnStatement",
      },
    ],
    "directives": [],
    "end": 205,
    "loc": SourceLocation {
      "end": Position {
        "column": 5,
        "index": 205,
        "line": 12,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 15,
        "index": 148,
        "line": 10,
      },
    },
    "start": 148,
    "type": "BlockStatement",
  },
  "computed": false,
  "end": 205,
  "generator": false,
  "id": null,
  "key": Node {
    "end": 145,
    "loc": SourceLocation {
      "end": Position {
        "column": 12,
        "index": 145,
        "line": 10,
      },
      "filename": undefined,
      "identifierName": "fullName",
      "start": Position {
        "column": 4,
        "index": 137,
        "line": 10,
      },
    },
    "name": "fullName",
    "start": 137,
    "type": "Identifier",
  },
  "kind": "method",
  "loc": SourceLocation {
    "end": Position {
      "column": 5,
      "index": 205,
      "line": 12,
    },
    "filename": undefined,
    "identifierName": undefined,
    "start": Position {
      "column": 4,
      "index": 137,
      "line": 10,
    },
  },
  "method": true,
  "params": [],
  "start": 137,
  "type": "ObjectMethod",
}
`;

exports[`sfc script parser should work options update data should work 1`] = `
"export default {
  name: 'AInput',
  data() {
    return {
      value: '1',
      number: 1
    };
  }
};"
`;

exports[`sfc script parser should work options update data should work 2`] = `
"export default {
  name: 'AInput',
  data() {
    return {
      value: true,
      number: 1
    };
  }
};"
`;

exports[`sfc script parser should work options update data should work 3`] = `
"export default {
  name: 'AInput',
  data() {
    return {
      number: 1
    };
  }
};"
`;

exports[`sfc script parser should work options update data should work 4`] = `
"export default {
  name: 'AInput',
  data() {
    return {};
  }
};"
`;

exports[`sfc script parser should work options update methods should work 1`] = `
"export default {
  name: 'AInput',
  methods: {
    handleChange(val) {
      this.value = val;
    },
    handleInput(val) {
      this.value = val;
    }
  }
};"
`;

exports[`sfc script parser should work options update methods should work 2`] = `
"export default {
  name: 'AInput',
  methods: {
    handleChange(val) {
      this.data = val;
    },
    handleInput(val) {
      this.value = val;
    }
  }
};"
`;

exports[`sfc script parser should work options update methods should work 3`] = `
"export default {
  name: 'AInput',
  methods: {
    handleInput(val) {
      this.value = val;
    }
  }
};"
`;

exports[`sfc script parser should work options update methods should work 4`] = `
{
  "async": false,
  "body": {
    "body": [
      {
        "expression": {
          "left": {
            "computed": false,
            "loc": undefined,
            "object": {
              "loc": undefined,
              "type": "ThisExpression",
            },
            "property": {
              "loc": undefined,
              "name": "value",
              "type": "Identifier",
            },
            "type": "MemberExpression",
          },
          "loc": undefined,
          "operator": "=",
          "right": {
            "loc": undefined,
            "name": "val",
            "type": "Identifier",
          },
          "type": "AssignmentExpression",
        },
        "loc": undefined,
        "type": "ExpressionStatement",
      },
    ],
    "directives": [],
    "type": "BlockStatement",
  },
  "computed": false,
  "generator": false,
  "key": {
    "name": "handleInput",
    "type": "Identifier",
  },
  "kind": "method",
  "params": [
    {
      "name": "val",
      "type": "Identifier",
    },
  ],
  "type": "ObjectMethod",
}
`;

exports[`sfc script parser should work options update watch should work 1`] = `
"export default {
  name: 'AInput',
  data() {
    return {
      firstName: 'Davide',
      lastName: 'Li',
      fullName: ''
    };
  },
  watch: {
    firstName(val) {
      this.fullName = val + this.lastName;
    },
    lastName(val) {
      this.fullName = this.firstName + ' ' + val;
    }
  }
};"
`;

exports[`sfc script parser should work options update watch should work 2`] = `
"export default {
  name: 'AInput',
  data() {
    return {
      firstName: 'Davide',
      lastName: 'Li',
      fullName: ''
    };
  },
  watch: {
    firstName(val) {
      this.fullName = val + ' ' + this.lastName;
    },
    lastName(val) {
      this.fullName = this.firstName + ' ' + val;
    }
  }
};"
`;

exports[`sfc script parser should work options update watch should work 3`] = `
"export default {
  name: 'AInput',
  data() {
    return {
      firstName: 'Davide',
      lastName: 'Li',
      fullName: ''
    };
  },
  watch: {
    lastName(val) {
      this.fullName = this.firstName + ' ' + val;
    }
  }
};"
`;

exports[`sfc script parser should work options update watch should work 4`] = `
{
  "async": false,
  "body": {
    "body": [
      {
        "expression": {
          "left": {
            "computed": false,
            "loc": undefined,
            "object": {
              "loc": undefined,
              "type": "ThisExpression",
            },
            "property": {
              "loc": undefined,
              "name": "fullName",
              "type": "Identifier",
            },
            "type": "MemberExpression",
          },
          "loc": undefined,
          "operator": "=",
          "right": {
            "left": {
              "left": {
                "computed": false,
                "loc": undefined,
                "object": {
                  "loc": undefined,
                  "type": "ThisExpression",
                },
                "property": {
                  "loc": undefined,
                  "name": "firstName",
                  "type": "Identifier",
                },
                "type": "MemberExpression",
              },
              "loc": undefined,
              "operator": "+",
              "right": {
                "extra": {
                  "raw": "' '",
                  "rawValue": " ",
                },
                "loc": undefined,
                "type": "StringLiteral",
                "value": " ",
              },
              "type": "BinaryExpression",
            },
            "loc": undefined,
            "operator": "+",
            "right": {
              "loc": undefined,
              "name": "val",
              "type": "Identifier",
            },
            "type": "BinaryExpression",
          },
          "type": "AssignmentExpression",
        },
        "loc": undefined,
        "type": "ExpressionStatement",
      },
    ],
    "directives": [],
    "type": "BlockStatement",
  },
  "computed": false,
  "generator": false,
  "key": {
    "name": "lastName",
    "type": "Identifier",
  },
  "kind": "method",
  "params": [
    {
      "name": "val",
      "type": "Identifier",
    },
  ],
  "type": "ObjectMethod",
}
`;
